libtiledbsoma/src/reindexer/reindexer.cc

```diff
@@ -79,18 +77,33 @@ void IntIndexer::map_locations(
     }
     auto hsize = kh_size(hash_);
     LOG_DEBUG(fmt::format("[Re-indexer] khash size = {}", hsize));
-    if (threads > 1) {
-        tiledb_thread_pool_ = std::make_unique<tiledbsoma::ThreadPool>(threads);
-    }
+
     LOG_DEBUG(
         fmt::format("[Re-indexer] Thread pool started and hash table created"));
+    context_ = context;
 }
 
 void IntIndexer::lookup(const int64_t* keys, int64_t* results, int size) {
     if (size == 0) {
         return;
     }
-    if (tiledb_thread_pool_ == nullptr) {  // When concurrency is 1
+    // Creating threadpool lazily when the first lookup happens
+    if (context_ != nullptr && context_->thread_pool() == nullptr) {
```

Comment:
this code doesn't look thread safe to me -- e.g., if there are two concurrent calls to `lookup()`, they will both attempt to create a thread pool and write to `context_`.  This needs to be a critical section.

TL;DR - contexts must be multi-thread safe

Reply:
Related:  it seems to me that the creation of the thread pool for a context should be code that lives _inside_ the SOMAContext C++ class, not outside of it.   @nguyenv - thoughts?

Reply:
oic - this is where the threadpool gets created.  Yes I agree that it makes more sense to be in the C++ code.

Reply:
and in a critical section, right?  (unless I grossly misunderstand, it is acceptable for multiple threads to all use the same SOMAContext much like tiledb.Ctx)

Reply:
Yes that's correct. `SOMAContext` should behave just like `tiledb::Context` which includes running multiple threads using the same context, so we should create the threadpool in a thread-safe manner.

Reply:
Great catch. The older version of lookup was thread safe but with shared context that's no longer the case. Moving to context makes sense. My goal here was merely lazily creation.